core/server/data/migrations/versions/5.0/2022-03-28-15-25-backfill-mrr-adjustments-for-offers.js

```diff
@@ -0,0 +1,101 @@
+            .where('o.duration', '=', 'forever')
+            .whereNotNull('s.offer_id')
+            .where('s.mrr', '!=', 0)
+            .orderBy('or.created_at', 'asc');
+
+        if (offerRedemptions.length === 0) {
+            logging.info('No Offers redeemed, skipping migration');
+            return;
+        } else {
+            logging.info(`Adjusting MRR for ${offerRedemptions.length} Offer Redemptions`);
+        }
+
+        const memberIds = uniq(offerRedemptions.map(redemption => redemption.member_id));
+
+        const mrrEvents = await knex
+            .select('*')
+            .from('members_paid_subscription_events')
+            .whereIn('member_id', memberIds);
+
+        function storeEventOnMemberId(storage, event) {
+            return {
+                ...storage,
+                [event.member_id]: storage[event.member_id] ? storage[event.member_id].concat(event) : [event]
+            };
+        }
+
+        const mrrEventsByMemberId = mrrEvents.reduce(storeEventOnMemberId, {});
+
+        const updatedEvents = [];
+
+        function calculateMRR(event, redemption) {
+            if (redemption && event.interval !== redemption.discount_interval) {
+                logging.error('Found invalid price & redemption pair');
+                return calculateMRR(event);
+            }
+
+            if (!redemption) {
+                return event.interval === 'year' ? event.amount / 12 : event.amount;
+            }
+
+            if (redemption.discount_type === 'percent') {
+                return calculateMRR({
+                    interval: event.interval,
+                    amount: event.amount * (100 - redemption.discount_amount) / 100
+                });
+            }
+
+            return calculateMRR({
+                interval: event.interval,
+                amount: event.amount - redemption.discount_amount
+            });
+        }
+
+        offerRedemptions.forEach((redemption) => {
+            const memberEvents = mrrEventsByMemberId[redemption.member_id];
+
+            if (memberEvents.length !== 1) {
+                return;
+            }
+
+            const firstEvent = memberEvents[0];
+
+            const mrr = calculateMRR(firstEvent, redemption);
+
+            updatedEvents.push({
+                id: firstEvent.id,
+                member_id: firstEvent.member_id,
+                from_plan: firstEvent.from_plan,
+                to_plan: firstEvent.to_plan,
+                currency: firstEvent.currency,
+                source: firstEvent.source,
+                created_at: firstEvent.created_at,
+                mrr_delta: mrr
+            });
+        });
+
+        if (updatedEvents.length === 0) {
+            return;
+        }
+
+        const idsToDelete = updatedEvents.map(event => event.id);
+
+        await knex('members_paid_subscription_events').whereIn('id', idsToDelete).del();
+        await knex('members_paid_subscription_events').insert(updatedEvents);
```

Comment:
Does it make more sense for this to be `.batchInsert`?

Reply:
Yup!